### DISCUSSION TITLE
integer variables / Bonmin

### DISCUSSION THREAD

**Discussion by agnusarev** [coin-or/Bonmin#20] [Q&A] [answered]
I am trying to use Bonmin to solve problem with nonlinier objective function, but C++ program run with status INFEAS. I can't understand how to use bonmin with integer variables. constrain g[0] is infeasible.
Maybe need to set some options.
Source code in attachment.
I am hope for your help.
Best regards, Alexander Gnusarev.

    // (C) Copyright Carnegie Mellon University 2006
    // All Rights Reserved.
    // This code is published under the Eclipse Public License.
    //
    // Authors :
    // P. Bonami, Carnegie Mellon University
    //
    // Date :  03/17/2006
    #include "MyTMINLP.hpp"
    #include "BonAmplInterface.hpp"
    
    bool 
    MyTMINLP::get_variables_types(Index n, VariableType* var_types)
    {
    	for (int i = 0; i < _problem.S; ++i){
    			var_types[i] = INTEGER;
    	}
      return true;
    }
    
    
    bool 
    MyTMINLP::get_variables_linearity(Index n, Ipopt::TNLP::LinearityType* var_types)
    {
    	for (int i = 0; i < _problem.S; ++i){
    			var_types[i] = Ipopt::TNLP::LINEAR;
    	}
      return true;
    }
    
    
    bool 
    MyTMINLP::get_constraints_linearity(Index m, Ipopt::TNLP::LinearityType* const_types)
    {
      assert (m==1);
      for (int i = 0; i < _problem.S + 1; ++i){
    	  const_types[i] = Ipopt::TNLP::LINEAR;
      }
    
      return true;
    }
    bool 
    MyTMINLP::get_nlp_info(Index& n, Index&m, Index& nnz_jac_g,
                           Index& nnz_h_lag, TNLP::IndexStyleEnum& index_style)
    {
      n = _problem.S;
      m = 1;
      index_style = TNLP::C_STYLE;
      return true;
    }
    
    bool 
    MyTMINLP::get_bounds_info(Index n, Number* x_l, Number* x_u,
                                Index m, Number* g_l, Number* g_u)
    {
    	assert(n==_problem.S);
    	assert(m==1);
      
    	for (int i = 0; i < _problem.S; ++i){
    			x_l[i] = 0;
    			x_u[i] = 3;
    	}
    	
      g_l[0] = 0;
      g_u[0] = _problem.B;
      
      return true;
    }
    
    bool 
    MyTMINLP::get_starting_point(Index n, bool init_x, Number* x,
                                 bool init_z, Number* z_L, Number* z_U,
                                 Index m, bool init_lambda,
                                 Number* lambda)
    {
    	assert(n==_problem.S);
    	assert(m==1);
      
      assert(init_x);
      assert(!init_lambda);
    
      	for (int i = 0; i < _problem.S; ++i){
    			x[i] = 0;
    	}
      return true;
    }
    
    bool 
    MyTMINLP::eval_f(Index n, const Number* x, bool new_x, Number& obj_value)
    {
      assert(n==_problem.S);
    	for (int i = 0; i < _problem.N; ++i) {
    		double s = 0.;
    		for (int j = 0; j < _problem.S; ++j) {
    				if (x[j] > 0){ 
    					s += _problem.K[i][_problem.index[j]][x[j] - 1];
    				}
    		}
    		obj_value -=  _problem.omega[i] * (1 - exp(-_problem.lambda * (s 
    		+ _problem.UC[i]))) * (s / (s 
    		+ _problem.UC[i]));
    	}
      return true;
    }
    
    bool
    MyTMINLP::eval_g(Index n, const Number* x, bool new_x, Index m, Number* g)
    {
    	assert(n==_problem.S);
    	assert(m==1);
    	
    	for (int j = 0; j < _problem.S; ++j) {
    			g[0] += x[j];
    	}
    
      return true;
    }
    
    void
    MyTMINLP::finalize_solution(TMINLP::SolverReturn status,
                                Index n, const Number* x, Number obj_value)
    {
      std::cout<<"Problem status: "<<status<<std::endl;
      std::cout<<"Objective value: "<<obj_value<<std::endl;
     if(printSol_ && x != NULL){
        std::cout<<"Solution:"<<std::endl;
        for(int i = 0 ; i < _problem.S ; i++){
          std::cout<<"x["<<i<<"] = "<<x[i];
          if(i < n-1) std::cout<<", ";}
        std::cout<<std::endl;
      }
    }
    
    bool
    MyTMINLP::eval_jac_g(Index n, const Number* x, bool new_x,
                         Index m, Index nnz_jac, Index* iRow, Index *jCol,
                         Number* values)
    {
    	assert(n==_problem.S);
        return true;
    }
    
    bool
    MyTMINLP::eval_h(Index n, const Number* x, bool new_x,
                     Number obj_factor, Index m, const Number* lambda,
                     bool new_lambda, Index nele_hess, Index* iRow,
                     Index* jCol, Number* values)
    {
    	assert(n==_problem.S);
    	assert(m==1);
    
      return true;
    }
    bool
    MyTMINLP::eval_grad_f(Index n, const Number* x, bool new_x, Number* grad_f)
    {
      assert(n==_problem.S);
      
      	for (int j = 0; j < _problem.S; ++j) {
    			grad_f[j] = -1.;
    	}
    
      return true;
    }

**Comment by svigerske** [comment, marked as answer]
Don't know. Try starting with `g[0] = 0.0;` in `eval_g`.
`eval_jac_g` seems to be missing an implementation.

**Reply by agnusarev** [reply]
I tried. this does not work.
is eval_jac_g required?

**Reply by svigerske** [reply]
yes

**Reply by agnusarev** [reply]
Thank you very much!!!
Everything works now.
